### ISSUE TITLE
Different transaction behavior for embedded memory/plocal vs remote

### ISSUE THREAD

**Issue by jlowenz** [orientechnologies/orientdb#5355]
Hi all, 

I've come across interesting behavior, and I'm not sure if I'm doing something incorrectly or if there is a bug (thinking the latter...) 

If I create two vertices (using the Java API), and set each with a link property referring to the opposite vertex, then commit and attempt to query the vertices I get two kinds of behavior, depending on how I'm connected to a database.

If I use the connection urls: "memory:test-db" or "plocal:/tmp/test-db" (or something similar), one of the two links is broken (i.e. null) after the query. However, if the connection url is "remote:localhost/test-db" (and I have a 2.1.5 server running), then everything works as expected: the links are correct with no nulls. 

I have a (not so minimal) working example at https://github.com/jlowenz/odb-mwe.git. It's in Clojure (because that's what I'm using for my project), but the [core.clj](https://github.com/jlowenz/odb-mwe/blob/master/src/odb_mwe/core.clj) and [core-test.clj](https://github.com/jlowenz/odb-mwe/blob/master/test/odb_mwe/core_test.clj) should be pretty straightforward to follow. Running the tests with `lein midje` shows the exact same code failing for memory and plocal, while the remote server connection succeeds. 

I'll also be happy to know if I'm doing something wrong, as this behavior has been tripping me up for a few days.

Thanks.

**Comment by jlowenz** [comment]
Below is the state of the plocal database after running the MWE referenced above. Note that somehow the temp ID has been stored for the link. This is not the case when connecting to the remote server. I.e. the name field correctly shows `#9:1`. Note I am using the `OrientGraphFactory` and then calling `getTx()`, wrapped in a try/catch with `commit()` and `shutdown()`.

```txt
Connecting to database [plocal:/tmp/odb_local] with user 'admin'...OK
orientdb {db=odb_local}> select * from V

----+----+------+----+-----
#   |@RID|@CLASS|name|other
----+----+------+----+-----
0   |#9:0|V     |A   |#9:-3
1   |#9:1|V     |B   |#9:0 
----+----+------+----+-----
```

**Comment by jlowenz** [comment]
I realize now the title of this issue is not particularly accurate: it really should be "Different transaction behavior for memory/plocal vs. remote." Since I wrote this, I have uncovered an additional instance of a broken update that does NOT work under memory/plocal, but is fine under remote. What is going on here? I will modify the MWE in an attempt to show this behavior as well.

**Comment by laa** [comment]
Hi @jlowenz sorry for late reply we will look at your issue during next 3 weeks.

**Comment by sukesh-soman** [comment]
Hello,

Not sure if this is related or not, but I am also seeing a strange difference in behavior when we create a link against a plocal & remote DB. 

I ran the below commands on a database created as plocal and got the error mentioned below. If I create another db of type remote, it works just fine.

Am I missing something?

> orientdb> create database plocal:mytestdb11 root root123 plocal
> Creating database [plocal:mytestdb11] using the storage type [plocal]...
> Database created successfully.
> 
> Current database is: plocal:mytestdb11
> orientdb {db=mytestdb11}> create class Post;
> 
> Class created successfully. Total classes in database now: 11
> orientdb {db=mytestdb11}> create property Post.id integer;
> Property created successfully with id=1
> 
> orientdb {db=mytestdb11}> create class comments;          
> Class created successfully. Total classes in database now: 12
> 
> orientdb {db=mytestdb11}> create property comments.id integer;
> Property created successfully with id=1
> 
> orientdb {db=mytestdb11}> create property comments.postid integer;
> Property created successfully with id=2
> 
> _orientdb {db=mytestdb11}> create link commentsonpost from comments.postid to post.id;
> [          
> Error: com.orientechnologies.orient.core.exception.OCommandExecutionException: Error on execution of command: sql.create link commentsonpost from comments.postid to post.id
> 
> Error: java.util.UnknownFormatConversionException: Conversion = '%'_

**Comment by laa** [comment]
@sukesh-soman it is SQL issue could you create separate ticket for  your case ?

**Comment by sukesh-soman** [comment]
Sure laa. Created #5869

**Comment by laa** [comment]
Hi @jlowenz , sorry but I do not understand a thing in clojure code.
Also I have tried to debug code but without success. I do not think there is a issue in ODB because this simple test case was tested inside of project and by many users. Could you provide alternative Java version of pseudo algorithm so I will understand your test ?

**Comment by jlowenz** [comment]
@laa, I will convert the test case to Java to ensure that the error still exists and make it easier to understand... it might take a day or so...

**Comment by laa** [comment]
@jlowenz Thank you very much for  your time.

**Comment by laa** [comment]
Hi, any updates ?

**Comment by jlowenz** [comment]
I have converted the Clojure to Java, but haven't generated the equivalent tests yet (my day job got in the way)... I will get to it asap.

**Comment by jlowenz** [comment]
Ok - I have finished the JUnit test for the Java code. It is in the same repository. You can run:
```bash
lein junit odb_mwe
```

and it should have two failures. Don't forget to start a server with the name "odb_remote".

**Comment by laa** [comment]
That is not transaction but serialization issue.  @tglman as far as I know you already have similar one , hope this test will allow to fix both of them. Assigned to @tglman

**Comment by jlowenz** [comment]
Ok, I'm confused. So OrientDB does serialization for plocal and memory connections (why)? If not, then are you saying the behavior for the remote connection is incorrect? The remote behavior is the one I consider correct and functional...

**Comment by laa** [comment]
We perform serialization of record in binary format to save them on disk. Order of serialization is broken so you have this issue.

**Comment by jlowenz** [comment]
@laa - thanks, that makes more sense!

**Comment by tglman** [comment]
hi @jlowenz,

I reproduced and fixed in 2.1.x, for 2.2 the bug was not there because some refactor on transaction commit, i added the test anyway, closing.
